File: shanghai/stock/shkvmrdatadown/ShangHaiMRKVDataDown.java
import net.sf.json.JSONException;
import net.sf.json.JSONObject;
import org.apache.commons.codec.binary.Hex;
import sun.misc.BASE64Decoder;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.logging.Logger;

/**
 * Created by liaochengming on 2015-12-14.
 * 此类是用来下载上海电信kv表中的数据
 */
public class ShangHaiMRKVDataDown {


    private static final String HMAC_SHA1_ALGORITHM = "HmacSHA1";

    /**
     * @param args args[0] 指定下载的时间（分钟）1表示0-4分钟。。。
     *             args[1] 指定下载的时间（小时）1表示1小时之前
     */
    public static void main(String[] args) {

        String yearToHour = getYearToHour(Integer.valueOf(args[1]));


        try {

            //合并文件，判断保存下载文件的文件夹下是否有上个小时的文件
            //如果有就合并
            MergedFiles.doMer();

            String token = getToken();

            String fileOver;
            File writer_file;

            fileOver = "/home/telecom/shdx/bin/shdx_kv_down/write_over.txt";
            writer_file = new File(fileOver);
            BufferedWriter writer = new BufferedWriter(new FileWriter(writer_file));
            int min_start = Integer.valueOf(args[0]);
            ExecutorService cachedThreadPool = Executors.newCachedThreadPool();

            //创建4个进程，一个进程跑一分钟的数据
            for (int i = (min_start - 1) * 4; i < min_start * 4; i++) {

                cachedThreadPool.execute(new Thread(new DoGetValue(yearToHour, getMS(i), i, token)));

            }

            cachedThreadPool.shutdown();

            while (true) {

                if (cachedThreadPool.isTerminated()) {

                    //休息随机时间
                    Thread.sleep((int) (Math.random() * 20 * 1000));

                    String isOver = readFirstLines(fileOver);

                    if (isOver.equals("false")) {

                        writer.write("true1");

                    } else {

                        int trueNum = Integer.valueOf(isOver.substring(4));
                        trueNum = trueNum + 1;
                        writer.write("true" + trueNum);

                    }

                    writer.flush();
                    break;

                }

                Thread.sleep(2000);

            }

            writer.close();

        } catch (Exception e) {

            e.printStackTrace();

        }


    }

    /**
     * 读文件一行数据
     *
     * @param path path 文件路径
     * @return 一行数据
     */
    private static String readFirstLines(String path) {

        File file = new File(path);

        if (file.exists()) {

            BufferedReader reader = null;

            try {

                reader = new BufferedReader(new FileReader(file));
                String tempString;

                // 只读一行
                tempString = reader.readLine();
                reader.close();

                if (null == tempString) {

                    return "false";
                }

                return tempString;

            } catch (IOException ioException) {

                ioException.printStackTrace();
                return "false";

            } finally {

                if (reader != null) {

                    try {

                        reader.close();

                    } catch (IOException ioException) {

                        ioException.printStackTrace();

                    }
                }
            }
        }

        return "false";
    }

    /**
     * 将数字格式化
     *
     * @param num 分钟和秒钟的数字
     * @return 固定格式的数字，如9改成09
     */
    private static String getMS(int num) {

        String newNum;

        if (num < 10) {

            newNum = "0" + num;

        } else {

            newNum = "" + num;

        }

        return newNum;
    }

    /**
     * 获取当前时间hour小时之前的时间字符串
     *
     * @param hour 固定小时，指定为多少小时之前
     * @return yyyyMMddHH格式的字符串
     */
    private static String getYearToHour(int hour) {

        Date oldTime = new Date(System.currentTimeMillis() - hour * 60 * 60 * 1000);
        SimpleDateFormat sdFormatter = new SimpleDateFormat("yyyyMMddHH");
        return sdFormatter.format(oldTime);

    }

    /**
     * 根据key获取value
     *
     * @param key   为kv数据表中的指定k，用来请求value
     * @param token 固定的请求数据时需要的参数
     * @return 请求到的数据
     * @throws Exception
     */
    private static String getValue(String key, String token) throws Exception {

        String getValueByKey = "kv/getValueByKey?token=" + token + "&table=kunyan_to_upload_inter_tab_mr&key=" + key;
        try {

            String value = doGet(getValueByKey);
            JSONObject jbKV = JSONObject.fromObject(value);
            String kv = jbKV.getString("result");

            if (!kv.equals("null")) {

                String valueBase64 = JSONObject.fromObject(kv).getString("value");
                return getFromBASE64(valueBase64);

            } else {

                return null;

            }

        } catch (JSONException jsonException) {

            jsonException.printStackTrace();

        }

        return "1";
    }

    /**
     * 获取电信的token的方法
     *
     * @return 获取的token字符串
     * @throws Exception
     */
    private static String getToken() throws Exception {

        String apiKey = KunYanUser.API_KEY;
        String userName = KunYanUser.KUN_YAN_USER_NAME;
        String password = KunYanUser.KUN_YAN_PASSWORD;
        String getToken = "getToken?apiKey=" + apiKey + "&" + "sign=" + sign(md5Encode(password), userName + apiKey);
        JSONObject jbToken = JSONObject.fromObject(doGet(getToken));
        return jbToken.getString("result");

    }

    /**
     * 解码Base64字符串
     *
     * @param string base64字符串
     * @return 解码后的字符串
     */
    private static String getFromBASE64(String string) {

        if (string == null) return null;

        BASE64Decoder decoder = new BASE64Decoder();

        try {

            byte[] bytes = decoder.decodeBuffer(string);
            return new String(bytes);

        } catch (Exception ioException) {

            ioException.printStackTrace();
            return null;

        }
    }


    /**
     * 此方法为http请求
     *
     * @param toDo 指定所需的请求
     * @return 请求到的数据
     * @throws Exception
     */
    private static String doGet(String toDo) throws Exception {

        URL localURL = new URL(KunYanUser.SHDX_KV_URL + toDo);
        URLConnection connection = localURL.openConnection();
        HttpURLConnection httpURLConnection = (HttpURLConnection) connection;

        httpURLConnection.setConnectTimeout(30000);
        httpURLConnection.setReadTimeout(30000);
        httpURLConnection.setRequestProperty("Connection", "keep-alive");
        httpURLConnection.setRequestProperty("Accept-Charset", "utf-8");
        httpURLConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");

        InputStream inputStream = null;
        InputStreamReader inputStreamReader = null;
        BufferedReader reader = null;
        StringBuilder resultBuilder = new StringBuilder();
        String tempLine;

        try {

            if (httpURLConnection.getResponseCode() >= 300) {

                throw new Exception("HTTP Request is not success, Response code is " + httpURLConnection.getResponseCode());

            }

            inputStream = httpURLConnection.getInputStream();
            inputStreamReader = new InputStreamReader(inputStream);
            reader = new BufferedReader(inputStreamReader);

            while ((tempLine = reader.readLine()) != null) {

                resultBuilder.append(tempLine);

            }
        } catch (IOException ioException) {

            ioException.printStackTrace();

        } finally {

            if (reader != null) {

                reader.close();

            }

            if (inputStreamReader != null) {

                inputStreamReader.close();

            }

            if (inputStream != null) {

                inputStream.close();

            }
        }

        return resultBuilder.toString();
    }

    /**
     * 获取http请求时所需的签名
     *
     * @param secretKey 经过md5加密处理的秘钥
     * @param data      用户名称和apiKey
     * @return 签名字符串
     * @throws Exception
     */
    private static String sign(String secretKey, String data) throws Exception {

        SecretKeySpec signingKey = new SecretKeySpec(secretKey.getBytes(), HMAC_SHA1_ALGORITHM);
        Mac mac = Mac.getInstance(HMAC_SHA1_ALGORITHM);
        mac.init(signingKey);
        byte[] rawHmac = mac.doFinal(data.getBytes());
        return Hex.encodeHexString(rawHmac);

    }

    /**
     * 对字符串做md5编码
     *
     * @param str 需要md5编码的字符串
     * @return 做了md5编码之后的字符串
     * @throws Exception
     */
    private static String md5Encode(String str) throws Exception {

        MessageDigest md5;

        try {

            md5 = MessageDigest.getInstance("MD5");

        } catch (Exception ioException) {

            ioException.printStackTrace();
            return "";
        }

        byte[] byteArray = str.getBytes("UTF-8");
        byte[] md5Bytes = md5.digest(byteArray);
        StringBuilder hexValue = new StringBuilder();

        for (byte md5Byte : md5Bytes) {

            int val = ((int) md5Byte) & 0xff;
            if (val < 16) {

                hexValue.append("0");

            }

            hexValue.append(Integer.toHexString(val));
        }

        return hexValue.toString();
    }


    //类部类，线程类
    private static class DoGetValue implements Runnable {

        private String min = "";
        private String sec = "";
        private String yearToHour = "";
        private String key = "";
        private BufferedWriter writer = null;
        private int num;
        private String token;

        //构造方法，初始化类的变量
        DoGetValue(String yearToHour, String min, int num, String token) {

            this.num = num;
            this.min = min;
            this.yearToHour = yearToHour;
            this.token = token;

        }

        @Override
        public void run() {

            String filesPath = "/home/liaochengming/shdx/kv_down_files/";
            File writerFile;
            String fileName = yearToHour + num + ".txt";
            writerFile = new File(filesPath + fileName);

            try {

                writer = new BufferedWriter(new FileWriter(writerFile));

                for (int i = 0; i < 60; i++) {

                    if (i < 10) {

                        sec = "0" + i;

                    } else {

                        sec = "" + i;

                    }

                    for (int k = 1; k < 2500; k++) {

                        key = this.yearToHour + this.min + this.sec + k;
                        String value = SHKVDown.getValue(key, token);

                        if (null == value) {

                            break;

                        } else if (value.equals("1")) {

                            Logger.getLogger("kv_down").info("kv down exception!");

                        } else {

                            writer.write(value);
                            writer.write("\r\n");
                            writer.flush();

                        }
                    }
                }

                writer.close();

            } catch (Exception ioException) {

                ioException.printStackTrace();

            }

        }
    }
}

File: shanghai/stock/shkvmrdatadown/ShangHaiMRKVDataDown.scala
package com.kunyan

import java.io.{BufferedReader, InputStream, InputStreamReader, PrintWriter}
import java.net.{HttpURLConnection, SocketTimeoutException, URL, URLConnection}
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.concurrent.Executors
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

import org.apache.commons.codec.binary.Hex
import org.json.{JSONException, JSONObject}
import sun.misc.BASE64Decoder

import scala.collection.mutable.ListBuffer
import scala.util.control.Breaks._

/**
 * Created by lcm on 2016/8/15.
 * 此类用来从上海KV数据库接收
 * 上海离线的股票搜索查看数据
 */
object ShangHaiMRKVDataDown {


  def main(args: Array[String]) {

    val token = getToken
    val yearToHour = getYearToHour(Integer.valueOf(args(0)))
    val pool = Executors.newFixedThreadPool(60)
    var dataList = new ListBuffer[String]
    val write = new PrintWriter(args(1) + yearToHour + ".txt")

    for (minInt <- 0 to 59) {

      val run = new Runnable() {
        @Override
        def run() {
          var minute = ""

          if (minInt < 10) {
            minute = "0" + minInt
          } else {
            minute = "" + minInt
          }

          val yearToMinute = yearToHour + minute
          val data = getOneMinuteData(yearToMinute, token)
          dataList = dataList ++: data

        }
      }

      pool.execute(run)
    }

    pool.shutdown()

    while (!pool.isTerminated) {
      Thread.sleep(5000)
    }

    for (index <- dataList.indices) {
      write.write(dataList(index) + "\n")
    }

    write.close()

  }

  /**
   * 获取一分钟数据
   * @param yearToMinute 字符串年月日时分 如200808080808:2008年8月8日8点8分
   * @param token 请求参数
   * @return 数据集
   */
  def getOneMinuteData(yearToMinute: String, token: String): ListBuffer[String] = {

    val data = new ListBuffer[String]
    var second = ""

    for (secInt <- 0 to 59) {

      if (secInt < 10) {
        second = "0" + secInt
      } else {
        second = "" + secInt
      }

      breakable {

        var index = 0
        var empty = 0

        while (index < 2500) {

          index = index + 1
          val key = yearToMinute + second + index
          val aData = getAData(key, token)

          if (aData != "") {
            data.+=(aData)
          } else {
            empty = empty + 1
          }

          if (empty > 3) {
            break()
          }
        }
      }
    }

    data
  }

  /**
   * 从KV表取一条数据
   * @param key 从KV取数据的key
   * @param token 发送http请求所需的参数
   * @return 取到的数据字符串
   */
  def getAData(key: String, token: String): String = {

    val url = "kv/getValueByKey?token=" + token + "&table=" + Parameter.KV_TABLE_NAME + "&key=" + key
    var aData = ""

    var back = doGet(url)

    while (back == "time out") {
      back = doGet(url)
    }

    if (back != "") {

      try {

        val jsonOBJ = new JSONObject(back)
        val result = jsonOBJ.getString("result")

        if (result != "null") {

          val value = new JSONObject(result).getString("value")
          aData = new String((new BASE64Decoder).decodeBuffer(value))

        }

      } catch {

        case jSONException: JSONException =>
          jSONException.printStackTrace()

      }
    }

    aData
  }

  /**
   * 获取当前时间hour小时之前的时间字符串
   *
   * @param hour 固定小时，指定为多少小时之前
   * @return yyyyMMddHH格式的字符串
   */
  def getYearToHour(hour: Int): String = {

    val oldTime: Date = new Date(System.currentTimeMillis - hour * 60 * 60 * 1000)
    val sdFormatter: SimpleDateFormat = new SimpleDateFormat("yyyyMMddHH")
    sdFormatter.format(oldTime)

  }

  /**
   * 获取电信的token的方法
   * @return 获取的token字符串
   */
  def getToken: String = {

    val apiKey: String = Parameter.API_KEY
    val userName: String = Parameter.KUN_YAN_USER_NAME
    val password: String = Parameter.KUN_YAN_PASSWORD

    val getToken: String = "getToken?apiKey=" + apiKey + "&" + "sign=" + sign(md5Encode(password), userName + apiKey)
    val jbToken: JSONObject = new JSONObject(doGet(getToken))

    jbToken.getString("result")

  }

  /**
   * 获取http请求时所需的签名
   *
   * @param secretKey 经过md5加密处理的秘钥
   * @param data      用户名称和apiKey
   * @return 签名字符串
   */
  def sign(secretKey: String, data: String): String = {

    val signingKey: SecretKeySpec = new SecretKeySpec(secretKey.getBytes, Parameter.HMAC_SHA1_ALGORITHM)
    val mac: Mac = Mac.getInstance(Parameter.HMAC_SHA1_ALGORITHM)
    mac.init(signingKey)
    val rawHmac: Array[Byte] = mac.doFinal(data.getBytes)

    Hex.encodeHexString(rawHmac)

  }

  /**
   * 对字符串做md5编码
   *
   * @param str 需要md5编码的字符串
   * @return 做了md5编码之后的字符串
   */
  private def md5Encode(str: String): String = {

    var md5: MessageDigest = null

    try {

      md5 = MessageDigest.getInstance("MD5")

    } catch {

      case ioException: Exception =>
        ioException.printStackTrace()
        return ""

    }

    val byteArray: Array[Byte] = str.getBytes("UTF-8")
    val md5Bytes: Array[Byte] = md5.digest(byteArray)
    val hexValue: StringBuilder = new StringBuilder

    for (md5Byte <- md5Bytes) {

      val `val`: Int = md5Byte.toInt & 0xff

      if (`val` < 16) {
        hexValue.append("0")
      }

      hexValue.append(Integer.toHexString(`val`))
    }

    hexValue.toString()
  }

  /**
   * 此方法为http请求
   *
   * @param url 指定所需的请求
   * @return 请求到的数据
   */
  def doGet(url: String): String = {

    val localURL: URL = new URL(Parameter.SH_KV_URL_HEAD + url)
    val connection: URLConnection = localURL.openConnection
    val httpURLConnection: HttpURLConnection = connection.asInstanceOf[HttpURLConnection]

    httpURLConnection.setConnectTimeout(3000)
    httpURLConnection.setReadTimeout(3000)
    httpURLConnection.setRequestProperty("Connection", "keep-alive")
    httpURLConnection.setRequestProperty("Accept-Charset", "utf-8")
    httpURLConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")

    var inputStream: InputStream = null
    var inputStreamReader: InputStreamReader = null
    var reader: BufferedReader = null
    val resultBuilder: StringBuilder = new StringBuilder
    var tempLine: String = null

    try {

      if (httpURLConnection.getResponseCode == 200) {

        inputStream = httpURLConnection.getInputStream
        inputStreamReader = new InputStreamReader(inputStream)
        reader = new BufferedReader(inputStreamReader)

        while ( {
          tempLine = reader.readLine
          tempLine
        } != null) {

          resultBuilder.append(tempLine)

        }
      }

    } catch {

      case timeOut: SocketTimeoutException =>
        resultBuilder.append("time out")
      case exception: Exception =>
        exception.printStackTrace()

    } finally {

      if (reader != null) {
        reader.close()
      }

      if (inputStreamReader != null) {
        inputStreamReader.close()
      }

      if (inputStream != null) {
        inputStream.close()
      }
    }

    resultBuilder.toString()
  }
}
